### PR TITLE
Fix compilation error [WIP]

### LINKED ISSUE #340: Failed compilation with -Wincompatible-pointer-types error
```
In file included from /home/tle/Work/openmohaa/code/qcommon/../script/../qcommon/listener.h:29,
                 from /home/tle/Work/openmohaa/code/qcommon/../script/scriptvariable.h:27,
                 from /home/tle/Work/openmohaa/code/qcommon/listener.cpp:25:
/home/tle/Work/openmohaa/code/qcommon/../script/../qcommon/con_arrayset.h: In member function ‘void con_arrayset<k, v>::resize(int) [with k = command_t; v = command_t]’:
/home/tle/Work/openmohaa/code/qcommon/../script/../qcommon/con_arrayset.h:195:9: warning: ‘void operator delete [](void*)’ called on pointer ‘*this.con_arrayset<command_t, command_t>::reverseTable’ with nonzero offset 8 [-Wfree-nonheap-object]
  195 |         delete[] oldReverseTable;
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
[ 84%] Building C object CMakeFiles/openmohaa.dir/code/qcommon/cm_test.c.o
[ 84%] Building CXX object CMakeFiles/openmohaa.dir/code/qcommon/cm_trace_lbd.cpp.o
[ 84%] Building C object CMakeFiles/openmohaa.dir/code/qcommon/cm_trace.c.o
[ 84%] Building CXX object CMakeFiles/openmohaa.dir/code/qcommon/cm_trace_obfuscation.cpp.o
/home/tle/Work/openmohaa/code/script/scriptvariable.cpp: In destructor ‘ScriptConstArrayHolder::~ScriptConstArrayHolder()’:
/home/tle/Work/openmohaa/code/script/scriptvariable.cpp:334:18: warning: ‘void operator delete [](void*, std::size_t)’ called on pointer ‘*this.ScriptConstArrayHolder::constArrayValue’ with nonzero offset 8 [-Wfree-nonheap-object]
  334 |         delete[] offset;
      |                  ^~~~~~
/home/tle/Work/openmohaa/code/sys/sys_unix.c: In function ‘Sys_ListFiles’:
/home/tle/Work/openmohaa/code/sys/sys_unix.c:418:55: warning: passing argument 3 of ‘Sys_ListFilteredFiles’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  418 |                 Sys_ListFilteredFiles( directory, "", filter, list, &nfiles );
      |                                                       ^~~~~~
/home/tle/Work/openmohaa/code/sys/sys_unix.c:340:71: note: expected ‘char *’ but argument is of type ‘const char *’
  340 | void Sys_ListFilteredFiles( const char *basedir, char *subdirs, char *filter, char **list, int *numfiles )
      |                                                                 ~~~~~~^~~~~~
[ 84%] Building C object CMakeFiles/openmohaa.dir/code/qcommon/cmd.c.o
[ 85%] Building CXX object CMakeFiles/openmohaa.dir/code/qcommon/common.cpp.o
[ 85%] Building C object CMakeFiles/openmohaa.dir/code/qcommon/crc.c.o
[ 85%] Building C object CMakeFiles/openmohaa.dir/code/qcommon/cvar.c.o
[ 85%] Building CXX object CMakeFiles/openmohaa.dir/code/qcommon/files.cpp.o
/home/tle/Work/openmohaa/code/server/sv_ccmds.c: In function ‘SV_ArchiveLevelFile’:
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1881:42: error: passing argument 2 of ‘ge->ReadLevel’ from incompatible pointer type [-Wincompatible-pointer-types]
 1881 |                 if (!ge->ReadLevel(name, &cls.savedCgameState, &cls.savedCgameStateSize)) {
      |                                          ^~~~~~~~~~~~~~~~~~~~
      |                                          |
      |                                          void **
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1881:42: note: expected ‘byte **’ {aka ‘unsigned char **’} but argument is of type ‘void **’
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1881:64: error: passing argument 3 of ‘ge->ReadLevel’ from incompatible pointer type [-Wincompatible-pointer-types]
 1881 |                 if (!ge->ReadLevel(name, &cls.savedCgameState, &cls.savedCgameStateSize)) {
      |                                                                ^~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                |
      |                                                                size_t * {aka long unsigned int *}
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1881:64: note: expected ‘int *’ but argument is of type ‘size_t *’ {aka ‘long unsigned int *’}
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1904:40: error: passing argument 3 of ‘ge->WriteLevel’ from incompatible pointer type [-Wincompatible-pointer-types]
 1904 |         ge->WriteLevel(name, autosave, &cls.savedCgameState, &cls.savedCgameStateSize);
      |                                        ^~~~~~~~~~~~~~~~~~~~
      |                                        |
      |                                        void **
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1904:40: note: expected ‘byte **’ {aka ‘unsigned char **’} but argument is of type ‘void **’
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1904:62: error: passing argument 4 of ‘ge->WriteLevel’ from incompatible pointer type [-Wincompatible-pointer-types]
 1904 |         ge->WriteLevel(name, autosave, &cls.savedCgameState, &cls.savedCgameStateSize);
      |                                                              ^~~~~~~~~~~~~~~~~~~~~~~~
      |                                                              |
      |                                                              size_t * {aka long unsigned int *}
/home/tle/Work/openmohaa/code/server/sv_ccmds.c:1904:62: note: expected ‘int *’ but argument is of type ‘size_t *’ {aka ‘long unsigned int *’}
make[2]: *** [CMakeFiles/openmohaa.dir/build.make:412: CMakeFiles/openmohaa.dir/code/server/sv_ccmds.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 85%] Building CXX object CMakeFiles/openmohaa.dir/code/qcommon/huffman.cpp.o
/home/tle/Work/openmohaa/code/sys/sys_main.c: In function ‘Sys_PIDFileName’:
/home/tle/Work/openmohaa/code/sys/sys_main.c:183:24: warning: return discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  183 |                 return va( "%s/%s/%s", homePath, gamedir, PID_FILENAME );
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 85%] Building C object CMakeFiles/openmohaa.dir/code/qcommon/md4.c.o
[ 85%] Building C object CMakeFiles/openmohaa.dir/code/qcommon/md5.c.o
/home/tle/Work/openmohaa/code/qcommon/cvar.c: In function ‘Cvar_CompleteCvarName’:
/home/tle/Work/openmohaa/code/qcommon/cvar.c:1609:43: warning: passing argument 1 of ‘Com_SkipTokens’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
 1609 |                 char *p = Com_SkipTokens( args, 1, " " );
      |                                           ^~~~
In file included from /home/tle/Work/openmohaa/code/qcommon/cvar.c:24:
/home/tle/Work/openmohaa/code/qcommon/q_shared.h:1171:29: note: expected ‘char *’ but argument is of type ‘const char *’
 1171 | char *Com_SkipTokens( char *s, int numTokens, const char *sep );
      |                       ~~~~~~^
/home/tle/Work/openmohaa/code/qcommon/files.cpp: In function ‘int check_filecase(char*)’:
/home/tle/Work/openmohaa/code/qcommon/files.cpp:548:15: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
  548 |     basedir = ".";
      |               ^~~
/home/tle/Work/openmohaa/code/qcommon/common.cpp: In function ‘void Com_Error(int, const char*, ...)’:
/home/tle/Work/openmohaa/code/qcommon/common.cpp:451:1: warning: ‘noreturn’ function does return
  451 | }
      | ^
/home/tle/Work/openmohaa/code/qcommon/common.cpp: In function ‘void Com_Quit_f()’:
/home/tle/Work/openmohaa/code/qcommon/common.cpp:478:1: warning: ‘noreturn’ function does return
  478 | }
      | ^
make[1]: *** [CMakeFiles/Makefile2:339: CMakeFiles/openmohaa.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
In file included from /home/tle/Work/openmohaa/code/qcommon/../script/../qcommon/listener.h:29,
                 from /home/tle/Work/openmohaa/code/qcommon/../script/scriptvariable.h:27,
                 from /home/tle/Work/openmohaa/code/qcommon/listener.cpp:25:
/home/tle/Work/openmohaa/code/qcommon/../script/../qcommon/con_arrayset.h: In member function ‘void con_arrayset<k, v>::resize(int) [with k = command_t; v = command_t]’:
/home/tle/Work/openmohaa/code/qcommon/../script/../qcommon/con_arrayset.h:195:9: warning: ‘void operator delete [](void*)’ called on pointer ‘*this.con_arrayset<command_t, command_t>::reverseTable’ with nonzero offset 8 [-Wfree-nonheap-object]
  195 |         delete[] oldReverseTable;
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
[ 85%] Linking CXX shared library gameppc64el.so
[ 85%] Built target fgame
make: *** [Makefile:136: all] Error 2
Exit 2
```

My machine spec is:

```
# Fedora 40 ppc64le
$ gcc --version
gcc (GCC) 14.2.1 20240801 (Red Hat 14.2.1-1)
```

It's seems to me this issue is impacted by the strict default settings of GCC. I will see if I could get a PR to address the issue.

### PR DESCRIPTION
**Pull request by runlevel5**
Fixes #340

### CODE REVIEW

**Comment by runlevel5** [comment]
@smallmodel thanks for the feedback, I will update the PR accordingly

**Comment by runlevel5** [comment]
@smallmodel the suggested changes now brings more regression:

```
/home/tle/Work/openmohaa/code/fgame/g_main.cpp:1265:36: error: invalid conversion from ‘byte*’ {aka ‘unsigned char*’} to ‘byte’ {aka ‘unsigned char’} [-fpermissive]
 1265 |                 *savedCgameState = (byte*)gi.Malloc(*savedCgameStateSize);
      |                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                    |
      |                                    byte* {aka unsigned char*}
/home/tle/Work/openmohaa/code/fgame/g_main.cpp:1267:36: warning: converting to non-pointer type ‘byte’ {aka ‘unsigned char’} from NULL [-Wconversion-null]
 1267 |                 *savedCgameState = NULL;
      |                                    ^~~~
/home/tle/Work/openmohaa/code/fgame/g_main.cpp:1270:24: error: invalid conversion from ‘byte’ {aka ‘unsigned char’} to ‘void*’ [-fpermissive]
 1270 |         arc.ArchiveRaw(*savedCgameState, *savedCgameStateSize);
      |                        ^~~~~~~~~~~~~~~~
      |                        |
      |                        byte {aka unsigned char}
In file included from /home/tle/Work/openmohaa/code/qcommon/../qcommon/script.h:33,
                 from /home/tle/Work/openmohaa/code/fgame/entity.h:51,
                 from /home/tle/Work/openmohaa/code/fgame/g_main.cpp:29:                 
 ```
 
 I am not familiar with the `gi.Malloc` so just want to seek your guidance before I refactor everything

**Comment by smallmodel** [comment]
The correct parameters are
```cpp
// g_main.h
qboolean G_ArchiveLevel(const char *filename, byte** savedCgameState, size_t *savedCgameStateSize, qboolean autosave, qboolean loading);
void G_WriteLevel(const char *filename, qboolean autosave, byte** savedCgameState, size_t* savedCgameStateSize);
qboolean G_ReadLevel(const char *filename, byte** savedCgameState, size_t* savedCgameStateSize);

//g_public.h
void (*WriteLevel)(const char *filename, qboolean autosave, byte** savedCgameState, size_t* savedCgameStateSize);
qboolean (*ReadLevel)(const char *filename, byte** savedCgameState, size_t* savedCgameStateSize);
```
`byte**` is correct, the pointer value of `savedCgameState` is both read/de-referenced and modified by `G_ArchiveLevel`

File: code/fgame/g_main.cpp [hunk 1]
```diff
@@ -1258,16 +1258,16 @@ qboolean G_ArchiveLevel(const char *filename, byte** savedCgameState, int *saved
...
+        num = (int)*savedCgameStateSize;
+        arc.ArchiveInteger(&num);
         if (!arc.Saving()) {
             if (*savedCgameStateSize) {
                 *savedCgameState = (byte*)gi.Malloc(*savedCgameStateSize);
             } else {
                 *savedCgameState = NULL;
             }
         }
-        arc.ArchiveRaw(*savedCgameState, *savedCgameStateSize);
```
Comment (smallmodel) [inline code/fgame/g_main.cpp:1270]: This was correct (de-reference to archive the array)

**Review by smallmodel — CHANGES_REQUESTED** [review]
Almost done

**Comment by runlevel5** [comment]
@smallmodel few warnings here and there but no more errors. The PR is ready for review

File: code/qcommon/cvar.c [hunk 2]
```diff
@@ -1606,7 +1606,7 @@ void Cvar_CompleteCvarName( const char *args, int argNum )
 	if( argNum == 2 )
 	{
 		// Skip "<cmd> "
-		char *p = Com_SkipTokens( args, 1, " " );
+		char *p = Com_SkipTokens((char *)args, 1, " " );
```
Comment (runlevel5) [inline code/qcommon/cvar.c:1609]: Address the `warning: return discards ‘const’ qualifier`

**Pull request merged**